File: javax/src/main/java/me/panpf/javax/util/range/DateRange.java
package me.panpf.javax.util.range;

import org.jetbrains.annotations.NotNull;

import java.util.Date;
import java.util.Iterator;
import java.util.NoSuchElementException;

public abstract class DateRange implements Iterable<Date>, ClosedRange<Date> {

    @NotNull
    private final Date start;
    @NotNull
    private final Date endInclusive;
    private final int step;

    public DateRange(@NotNull Date start, @NotNull Date endInclusive, int step) {
        super();
        this.start = start;
        this.endInclusive = endInclusive;
        this.step = step;
    }

    @NotNull
    public Iterator<Date> iterator() {
        return (new DateRange.IteratorInternal(this, this.getStart(), this.getEndInclusive(), this.step));
    }

    public boolean contains(@NotNull Date value) {
        return this.step > 0 ? value.compareTo(this.getStart()) >= 0 && value.compareTo(this.getEndInclusive()) <= 0 : (this.step < 0 && (value.compareTo(this.getStart()) <= 0 && value.compareTo(this.getEndInclusive()) >= 0));
    }

    public boolean isEmpty() {
        return this.step > 0 ? this.getStart().compareTo(this.getEndInclusive()) > 0 : (this.step >= 0 || this.getStart().compareTo(this.getEndInclusive()) < 0);
    }

    @NotNull
    public abstract Date nextDate(@NotNull Date var1);

    @NotNull
    public Date getStart() {
        return this.start;
    }

    @NotNull
    public Date getEndInclusive() {
        return this.endInclusive;
    }

    public final int getStep() {
        return this.step;
    }

    public static final class IteratorInternal implements Iterator<Date> {
        private boolean hasNext;
        @NotNull
        private Date next;
        private final DateRange range;
        private final Date last;
        private final int step;

        IteratorInternal(@NotNull DateRange range, @NotNull Date first, @NotNull Date last, int step) {
            super();
            this.range = range;
            this.last = last;
            this.step = step;
            this.hasNext = this.step > 0 ? first.compareTo(this.last) <= 0 : (this.step < 0 && first.compareTo(this.last) >= 0);
            this.next = this.hasNext ? first : this.last;
        }

        public boolean hasNext() {
            return this.hasNext;
        }

        @NotNull
        public Date next() {
            if (!this.hasNext) {
                throw (new NoSuchElementException());
            } else {
                Date result = this.next;
                this.next = this.range.nextDate(this.next);
                this.hasNext = this.step > 0 ? this.next.compareTo(this.last) <= 0 : (this.step < 0 && this.next.compareTo(this.last) >= 0);
                return result;
            }
        }

        public void remove() {
            throw new UnsupportedOperationException("Operation is not supported for read-only collection");
        }
    }
}

File: javax-kt/src/main/java/me/panpf/javaxkt/util/DateRange.kt
/*
 * Copyright (C) 2018 Peng fei Pan <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.panpf.javaxkt.util

import java.util.*

abstract class DateRange(override val start: Date, override val endInclusive: Date, val step: Int) : Iterable<Date>, ClosedRange<Date> {

    override fun iterator(): Iterator<Date> = IteratorInternal(this, start, endInclusive, step)

    override fun contains(value: Date): Boolean {
        @Suppress("ConvertTwoComparisonsToRangeCheck")
        return when {
            step > 0 -> value >= start && value <= endInclusive
            step < 0 -> value <= start && value >= endInclusive
            else -> false
        }
    }

    override fun isEmpty(): Boolean {
        return when {
            step > 0 -> start > endInclusive
            step < 0 -> start < endInclusive
            else -> true
        }
    }

    abstract fun nextDate(date: Date): Date

    class IteratorInternal(private val range: DateRange, first: Date, private val last: Date, private val step: Int) : Iterator<Date> {
        var hasNext: Boolean
        var next: Date

        init {
            hasNext = when {
                step > 0 -> first <= last
                step < 0 -> first >= last
                else -> false
            }
            next = if (hasNext) first else last
        }

        override fun hasNext(): Boolean {
            return hasNext
        }

        override fun next(): Date {
            if (!hasNext) {
                throw NoSuchElementException()
            }

            val result = next
            next = range.nextDate(next)
            hasNext = when {
                step > 0 -> next <= last
                step < 0 -> next >= last
                else -> false
            }
            return result
        }
    }
}

class YearRange(start: Date, endInclusive: Date, step: Int) : DateRange(start, endInclusive, step) {
    override fun nextDate(date: Date): Date = date.addYear(step)
}

class MonthRange(start: Date, endInclusive: Date, step: Int) : DateRange(start, endInclusive, step) {
    override fun nextDate(date: Date): Date = date.addMonth(step)
}

class DayRange(start: Date, endInclusive: Date, step: Int) : DateRange(start, endInclusive, step) {
    override fun nextDate(date: Date): Date = date.addDayOfMonth(step)
}

class HourRange(start: Date, endInclusive: Date, step: Int) : DateRange(start, endInclusive, step) {
    override fun nextDate(date: Date): Date = date.addHourOfDay(step)
}

class MinuteRange(start: Date, endInclusive: Date, step: Int) : DateRange(start, endInclusive, step) {

    override fun nextDate(date: Date): Date = date.addMinute(step)
}
class SecondRange(start: Date, endInclusive: Date, step: Int) : DateRange(start, endInclusive, step) {

    override fun nextDate(date: Date): Date = date.addSecond(step)
}
class MillisecondRange(start: Date, endInclusive: Date, step: Int) : DateRange(start, endInclusive, step) {
    override fun nextDate(date: Date): Date = date.addMillisecond(step)
}

